crates/ironrdp-pdu/src/rdp.rs

```diff
@@ -28,34 +29,70 @@ pub struct ClientInfoPdu {
     pub client_info: ClientInfo,
 }
 
+impl ClientInfoPdu {
+    const NAME: &'static str = "ClientInfoPDU";
+    const FIXED_PART_SIZE: usize = BasicSecurityHeader::FIXED_PART_SIZE + ClientInfo::FIXED_PART_SIZE;
+    const MAX_SIZE: usize = BasicSecurityHeader::FIXED_PART_SIZE + ClientInfo::MAX_SIZE;
+}
+
+impl PduEncode for ClientInfoPdu {
+    fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
+        ensure_fixed_part_size!(in: dst);
+
+        self.security_header.encode(dst)?;
+        self.client_info.encode(dst)?;
+
+        Ok(())
+    }
+
+    fn name(&self) -> &'static str {
+        Self::NAME
+    }
+
+    fn size(&self) -> usize {
+        self.security_header.size() + self.client_info.size()
+    }
+}
+
+impl<'de> PduDecode<'de> for ClientInfoPdu {
+    fn decode(src: &mut ReadCursor<'de>) -> PduResult<Self> {
+        ensure_fixed_part_size!(in: src);
+
+        let security_header = BasicSecurityHeader::decode(src)?;
+        if !security_header.flags.contains(BasicSecurityHeaderFlags::INFO_PKT) {
+            return Err(invalid_message_err!("securityHeader", "got invalid security header"));
+        }
+
+        let client_info = ClientInfo::decode(src)?;
+
+        Ok(Self {
+            security_header,
+            client_info,
+        })
+    }
+}
+
 impl PduParsing for ClientInfoPdu {
     type Error = RdpError;
 
     fn from_buffer(mut stream: impl io::Read) -> Result<Self, Self::Error> {
-        let security_header = BasicSecurityHeader::from_buffer(&mut stream)?;
-        if security_header.flags.contains(BasicSecurityHeaderFlags::INFO_PKT) {
-            let client_info = ClientInfo::from_buffer(&mut stream)?;
-
-            Ok(Self {
-                security_header,
-                client_info,
-            })
-        } else {
-            Err(RdpError::InvalidPdu(String::from(
-                "Expected ClientInfo PDU, got invalid SecurityHeader flags",
-            )))
-        }
+        let mut buf = [0; Self::MAX_SIZE];
+        let len = stream.read(&mut buf)?;
+        let mut cursor = ReadCursor::new(&buf[..len]);
+        Ok(Self::decode(&mut cursor).unwrap())
     }
 
     fn to_buffer(&self, mut stream: impl io::Write) -> Result<(), Self::Error> {
-        self.security_header.to_buffer(&mut stream)?;
-        self.client_info.to_buffer(&mut stream)?;
-
+        let mut buf = [0; Self::MAX_SIZE];
+        let mut cursor = WriteCursor::new(&mut buf);
+        Self::encode(self, &mut cursor).unwrap();
+        let len = cursor.pos();
+        stream.write_all(&buf[..len]).unwrap();
         Ok(())
     }
 
     fn buffer_length(&self) -> usize {
-        self.security_header.buffer_length() + self.client_info.buffer_length()
+        self.size()
     }
 }
```

Comment:
This could probably be a blanket implementation for all types implementing `PduEncode` and `PduDecode` :+1: 
We would need to remove the `unwrap`s however, but we can probably just set `type Error = PduError`?

Reply:
I think it’s probably best to take the big `MAX_SIZE` approach, because even if it’s not great, we can keep the "observable" behavior unchanged, which is very important at least for bisecting regressions as you mention. And if it’s only happening during the handshake sequence, it’s not too bad either.

Also, we could create a branch in which we accumulate all the incremental changes until we have one "big PR" changing everything, but properly reviewed via intermediate smaller PRs :thinking: 
Would you be open to that?

Reply:
Ok, that sounds good. Give me a few days to rework my changes and publish. Thanks!